docs/source/generics.rst

```diff
@@ -0,0 +1,984 @@
+
+Other uses are factory methods, such as copy and deserialization.
+For class methods, you can also define generic ``cls``, using :py:class:`Type[T] <typing.Type>`:
+
+.. code-block:: python
+
+   from typing import TypeVar, Type
+
+   T = TypeVar('T', bound='Friend')
+
+   class Friend:
+       other: "Friend" = None
+
+       @classmethod
+       def make_pair(cls: Type[T]) -> tuple[T, T]:
+           a, b = cls(), cls()
+           a.other = b
+           b.other = a
+           return a, b
+
+   class SuperFriend(Friend):
+       pass
+
+   a, b = SuperFriend.make_pair()
+
+Note that when overriding a method with generic ``self``, you must either
+return a generic ``self`` too, or return an instance of the current class.
+In the latter case, you must implement this method in all future subclasses.
+
+Note also that the type checker may not always verify that the implementation of a copy
+or a deserialization method returns the actual type of self. Therefore
+you may need to silence the type checker inside these methods (but not at the call site),
+possibly by making use of the ``Any`` type or a ``# type: ignore`` comment.
+
+For some advanced uses of self types, see :ref:`additional examples <advanced_self>`.
+
+Automatic self types using typing.Self
+**************************************
+
+Since the patterns described above are quite common, a simpler syntax
+was introduced in :pep:`673`.
+
+Instead of defining a type variable and using an explicit annotation
+for ``self``, you can use the special type ``typing.Self``. This is
+automatically transformed into a type variable with the current class
+as the upper bound, and you don't need an annotation for ``self`` (or
+``cls`` in class methods).
+
+Here's what the example from the previous section looks like
+when using ``typing.Self``:
+
+.. code-block:: python
+
+   from typing import Self
+
+   class Friend:
+       other: Self | None = None
+
+       @classmethod
+       def make_pair(cls) -> tuple[Self, Self]:
+           a, b = cls(), cls()
+           a.other = b
+           b.other = a
+           return a, b
+
+   class SuperFriend(Friend):
+       pass
+
+   a, b = SuperFriend.make_pair()
+
+This is more compact than using explicit type variables. Also, you can
+use ``Self`` in attribute annotations in addition to methods.
+
+.. note::
+
+   To use this feature on Python versions earlier than 3.11, you will need to
+   import ``Self`` from ``typing_extensions`` (version 4.0 or newer).
+
+.. _variance-of-generics:
+
+Variance of generic types
+*************************
+
+There are three main kinds of generic types with respect to subtype
+relations between them: invariant, covariant, and contravariant.
+Assuming that we have a pair of types ``Animal`` and ``Bear``, and
+``Bear`` is a subtype of ``A``, these are defined as follows:
+
+* A generic class ``MyCovGen[T]`` is called covariant in type variable
+  ``T`` if ``MyCovGen[Bear]`` is always a subtype of ``MyCovGen[Animal]``.
```

Comment:
I think leaving out "always" here and below would make the wording more clear.

Reply:
Good point, fixed